lib/elixir/lib/file.ex

```diff
@@ -791,35 +795,38 @@ defmodule File do
   Returns `:ok` otherwise.
   """
   @spec cp!(Path.t(), Path.t(), (Path.t(), Path.t() -> boolean)) :: :ok
-  def cp!(source, destination, callback \\ fn _, _ -> true end) do
-    case cp(source, destination, callback) do
+  def cp!(source_file, destination_file, callback \\ fn _, _ -> true end) do
+    case cp(source_file, destination_file, callback) do
       :ok ->
         :ok
 
       {:error, reason} ->
         raise File.CopyError,
           reason: reason,
           action: "copy",
-          source: IO.chardata_to_string(source),
-          destination: IO.chardata_to_string(destination)
+          source: IO.chardata_to_string(source_file),
+          destination: IO.chardata_to_string(destination_file)
     end
   end
 
   @doc ~S"""
-  Copies the contents in source to destination.
+  Copies the contents in `source` to `destination` recursively, maintaining the
+  source directory structure and modes.
+
+  If `source` is a file or a symbolic link to it, `destination` must be a path
+  to an existent file, a sysmbolic link to one, or a path to a non-existent file.
+
+  If `source` is a directory, or a symbolic link to it, then `destination` must
+  be an existent `directory` or a symbolic link to one, or a path to a non-existent directory.
 
   If the source is a file, it copies `source` to
-  `destination`. If the source is a directory, it copies
-  the contents inside source into the destination.
+  `destination`. If the `source` is a directory, it copies
+  the contents inside source into the `destination` directory.
 
   If a file already exists in the destination, it invokes `callback`.
   `callback` must be a function that takes two arguments: `source` and `destination`.
   The callback should return `true` if the existing file should be overwritten and `false` otherwise.
 
-  If a directory already exists in the destination
-  where a file is meant to be (or vice versa), this
-  function will fail.
```

Comment:
I'm still not able to understand what the "vice versa" translates back to plain simple English,
but in any case the information about the directory existing in the destination and the function failing is wrong.

Reply:
If there is a file in the source and a directory in the destination OR if there is a directory in the source and a file in the destination, it fails.

Reply:
Personally it feels like this is clear enough (not a native speaker). Should we put it back?

Reply:
We can expand it without the vice-versa. Being precise doesn't hurt. :)

Reply:
It think we can still remove it, because it is covered by the two opening paragraphs

> 
>   If `source` is a file or a symbolic link to it, `destination` must be a path
>   to an existent file, a symbolic link to one, or a path to a non-existent file.
> 
>   If `source` is a directory, or a symbolic link to it, then `destination` must
>   be an existent `directory` or a symbolic link to one, or a path to a non-existent directory.
>